ttnn/cpp/ttnn/operations/experimental/ccl/llama_reduce_scatter_create_heads/device/kernels/dataflow/writer_llama_reduce_scatter.cpp

```diff
@@ -0,0 +1,217 @@
+            ? num_pages_per_packet
+            : input_shard_cores_per_device * tiles_per_core_width % num_pages_per_packet;
+    constexpr size_t packet_header_size = sizeof(PACKET_HEADER_TYPE);
+
+    // Precomputed constants for better optimization
+    constexpr uint32_t chip_id_offset = chip_id * num_pages_per_packet * page_size_bytes;
+
+    constexpr uint32_t other_devices = num_devices - 1;
+
+    // Runtime arguments
+    uint32_t receiver_semaphore_address = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t local_semaphore_address = get_semaphore(get_arg_val<uint32_t>(rt_arg_idx++));
+    bool sender_core = (bool)get_arg_val<uint32_t>(rt_arg_idx++);
+    bool worker_core = (bool)get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t linear_output_page_start_idx = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t sender_packet_start = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t sender_packet_end = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t sender_total_num_pages = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t q_base_addr = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t k_base_addr = get_arg_val<uint32_t>(rt_arg_idx++);
+    uint32_t v_base_addr = get_arg_val<uint32_t>(rt_arg_idx++);
+
+    if (sender_core) {
+        auto fabric_connection =
+            FabricConnectionManager::build_from_args<FabricConnectionManager::BUILD_AND_OPEN_CONNECTION_START_ONLY>(
+                rt_arg_idx);
+        // Set up packet headers once
+        constexpr uint8_t device_order[other_devices] =
+            DEVICE_ORDER;  // this is code gen'd in the program factory using the defines
+        constexpr uint8_t packet_worker_cores[num_packet_worker_cores][2] = PACKET_WORKER_CORES;
+        const auto packet_header_buffer_addr = get_read_ptr(packet_header_cb_id);
+        auto* unicast_packet_header = reinterpret_cast<volatile PACKET_HEADER_TYPE*>(packet_header_buffer_addr);
+        auto* sem_inc_packet_header =
+            reinterpret_cast<volatile PACKET_HEADER_TYPE*>(packet_header_buffer_addr + packet_header_size);
+        const uint64_t sem_noc_addr =
+            get_noc_addr(packet_receiver_core_x, packet_receiver_core_y, receiver_semaphore_address);
+        sem_inc_packet_header->to_noc_unicast_atomic_inc(tt::tt_fabric::NocUnicastAtomicIncCommandHeader{
+            sem_noc_addr,
+            static_cast<uint16_t>(1),  // increment 1
+            32});
+
+        const uint32_t base_receiver_l1_addr = get_read_ptr(fabric_receiver_cb_id);
+
+        // Precompute the packet offset once
+        const uint32_t packet_offset = base_receiver_l1_addr + chip_id_offset;
+
+        if (fabric_connection.is_logically_connected()) {
+            fabric_connection.open_finish();
+        }
+        for (uint32_t target_device_id : device_order) {
+            // Calculate device-specific constants once per device
+            bool forward_connection = true;
+            if (RING_TOPOLOGY) {
```

Comment:
I assume this can be 
```suggestion
            if constexpr (RING_TOPOLOGY) {
```
since it shouldn't change by worker or op invocation?